### PR TITLE
Use `wal_generate` in tests to test complex WAL situations

### LINKED ISSUE #1190: Confusion about end LSNs of an XLOG_SWITCH record
While I was looking at the code around WAL decoding, and which LSNs are record start LSNs and which ones are end LSNs, and where we restart replication (https://github.com/zenithdb/zenith/pull/1182), I noticed an issue with XLOG_SWITCH records:

When `WalStreamDecoder::poll_decode` has successfully decoded a record, it returns its end-LSN to the caller. After the record is processed in the pageserver, that end LSN become the current `last_record_lsn`, and if a checkpoint happens, it is stored as the `disk_consistent_lsn`. After restart, we will restart replication from the safekeepers at that LSN.

However, if the last record was an XLOG_SWITCH record, that won't work. An XLOG_SWITCH record is special, in that it "consumes" the rest of the WAL segment, so that the next record will start at the beginning of the next segment. However, for an XLOG_SWITCH record, `poll_decode` returns the position of the end of the record, not the end of the segment. If you try to restart replication from the end of the record, I believe it will fail.

Safekeeper uses the same code, so there might be issues there too.

I haven't tested that, but it seems wrong to me. Needs to be verified and fixed if necessary.

### PR DESCRIPTION
**Pull request by yeputons**
This PR uses the `postgres_ffi/wal_generate` binary to add more end-to-end tests. This binary can generate complicated WALs, which may trigger some edge cases in WAL decoding code, like #1190 or #544.

It may be easier to review with `git show HEAD --color-moved`

Follow-up PR will expose and fix #1190.

### CODE REVIEW

File: test_runner/batch_others/test_wal_generate.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    log.info("Restarting all safekeepers and pageservers")
+    env.pageserver.stop()
+    env.safekeepers[0].stop()
+    env.safekeepers[0].start()
+    env.pageserver.start()
+
+    log.info("Trying more queries")
+    res = pg.safe_psql_many(queries=[
+        'SELECT SUM(key) FROM keys',
+        'INSERT INTO keys SELECT generate_series(101, 200)',
```
Comment (yeputons) [inline test_runner/batch_others/test_wal_generate.py:41]: For some reason this query takes about 10 seconds on my machine when `wal_type != 'simple'`. Any ideas? Decoding 16 MB should not take that long.

**Comment by yeputons** [comment]
All failures seem to come from #1985 (bug) or #1966 (not merged into my HEAD branch yet). Will rebase on `main` once it becomes green.

**Comment by yeputons** [comment]
After discussion with @arssher: I consider this to be an important end-to-end test. Unit tests for individual WAL parsing pieces are good as well, but in the end we worry about the user generating a weird WAL in our system as a whole.

File: test_runner/batch_others/test_wal_generate.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
+from fixtures.neon_fixtures import NeonEnvBuilder, WalGenerate
+from fixtures.log_helper import log
+import pytest
+
+# Ensure that complicated WALs are correctly handled by the whole stack.
```
Comment (arssher) [inline test_runner/batch_others/test_wal_generate.py:5]:
```suggestion
# Restart storage nodes with WAL end having specially crafted shape, like last record crossing segment boundary, to test decoding issues.
```

**Review by arssher — APPROVED** [review]

**Comment by arssher** [comment]
Hm. Why this test currently passes with `last_wal_record_crossing_segment` variant -- safekeepers part is not fixed yet?

Also, I'd rename test and file name to something like `test_crafted_wal_end`.

FWIW I feel we'd be ok with just python tests. Separate WAL generator binary and wrapper around it looks too much for this, but anyway, we already have them.

**Comment by arssher** [comment]
Ah, ok.
